openchs-server-api/src/main/java/org/openchs/service/ProgramEncounterService.java

```diff
@@ -1,19 +1,47 @@
 package org.openchs.service;
 
+import com.bugsnag.Bugsnag;
+import org.openchs.dao.EncounterTypeRepository;
 import org.openchs.dao.ProgramEncounterRepository;
-import org.openchs.domain.ProgramEncounter;
+import org.openchs.dao.ProgramEnrolmentRepository;
+import org.openchs.dao.individualRelationship.RuleFailureLogRepository;
+import org.openchs.domain.*;
+import org.openchs.geo.Point;
+import org.openchs.web.AbstractController;
+import org.openchs.web.IndividualController;
 import org.openchs.web.request.EncounterTypeContract;
+import org.openchs.web.request.PointRequest;
+import org.openchs.web.request.ProgramEncounterRequest;
 import org.openchs.web.request.ProgramEncountersContract;
+import org.openchs.web.request.rules.RulesContractWrapper.VisitSchedule;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Service;
 
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
 @Service
-public class ProgramEncounterService {
+public class ProgramEncounterService extends AbstractController<ProgramEncounter> {
```

Comment:
Service should not be extending controller. I think you are doing this because you want to use newOrExistingEntity method. Extract out this method in EntityHelper class. So then you can use static EntityHelper.newOrExistingEntity method in here. Put EntityHelper class in common or framework package. And also change AbstractController to internally call EntityHelper.newOrExistingEntity so there is no duplication.